docs/proposals/kubegen.md

```diff
@@ -0,0 +1,214 @@
+# Kubernetes Code Generators
+
+Kubernetes provides a collection of code generators to assist with developing APIs.  These code generators generate
+useful libraries such as clients, openapi definitions, copying go structs, etc.
+
+The `kubegen` uses directory structure conventions to detect Kubernetes APIs definitions and run the standard
+set of Kubernetes code generators.
+
+There are 2 methods for running kubegen
+
+- Directly through the kubegen command line by downloading the binary and running at a the project root
+- Through Bazel by adding rules to `WORKSPACE` and `BUILD.gazel`
+
+## Kubernetes API conventions
+
+kubegen will run the full set of Kubernetes code generators for all APIs it finds in the project.
+kubegen looks for API groups defined under `pkg/apis/<group>/<version>` where *group* and *version* match the following
+patterns.
+
+- **group pattern**: ^[a-z]+$
+- **version pattern**: ^v\\d+(alpha\\d+|beta\\d+)*$
+
+By default, kubegen will run code generators for both external types defined under `pkg/apis/<group>/<version>` and
+internal types defined under `pkg/apis/<group>`.  The location kubegen searches can be overridden.
+
+kubegen will prepend all generated files with copyright and license comments.
+
+By default, the kubegen binary runs all* of the Kubernetes code
```

Comment:
what is all? The list varies depending on external-type-only projects (usually CRDs of k8s.io/api like repos) and user-provider APIServers.

Reply:
the list is directly below this statement.  crd use cases can manually specify the generators they need

Reply:
Don't like to specify the generators manually. We have these two pretty well defined use-cases: external types only, internal+external. They should just work out of the box. If things are not needed (e.g. protobuf), this should be auto-detected by the lack of tags.

Reply:
I think you are right.  We may need some break glass in the future, but best not to start with it.

Are there existing tags we should us?

Reply:
Does this allow us to disable things like deepcopy and conversiongen?